keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,216 @@
+---
+title: cgroups v2
+authors:
+  - "@giuseppe"
+owning-sig: sig-node
+participating-sigs:
+      - sig-architecture
+reviewers:
+  - "@yujuhong"
+  - "@dchen1107"
+  - "@derekwaynecarr"
+approvers:
+  - "@yujuhong"
+  - "@dchen1107"
+  - "@derekwaynecarr"
+editor: Giuseppe Scrivano
+creation-date: 2019-11-18
+last-updated: 2019-11-18
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Cgroups v2
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [User Stories](#user-stories)
+- [Implementation Details](#implementation-details)
+  - [Dependencies on OCI and container runtimes](#implementation-details)
+  - [Current status of dependencies](#current-status-of-dependencies)
+  - [Current cgroups usage and the equivalent in cgroups v2](#current-cgroups-usage-and-the-equivalent-in-cgroups-v2)
+- [Proposal](#proposal)
+  - [Option 1: Convert from cgroups v1 settings to v2](#option-1-convert-from-cgroups-v1-settings-to-v2)
+  - [Option 2: Use cgroups v2 throughout the stack](#option-2-use-cgroups-v2-throughout-the-stack)
+- [Risk and Mitigations](#risk-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing](#testing)
+
+<!-- /toc -->
+
+## Summary
+
+A proposal to add support for cgroups v2 to kubernetes.
+
+## Motivation
+
+The new kernel cgroups v2 API was declared stable more than two years
+ago. Newer features in the kernel such as PSI depend upon cgroups
+v2. groups v1 will eventually become obsolete in favor of cgroups v2.
+Some distros are already using cgroups v2 by default, and that
+prevents Kubernetes from working as it is required to run with cgroups
+v1.
+
+## Goals
+
+This proposal aims to:
+
+*   Add support for cgroups v2 to cAdvisor ([https://github.com/google/cadvisor/pull/2309](https://github.com/google/cadvisor/pull/2309))
+*   Add support for cgroups v2 to the Kubelet
+
+## User Stories
+
+*   The Kubelet can run on a host using either cgroups v1 or v2
+*   Scheduler should be able to use cgroups v2
+
+## Implementation Details
+
+### Dependencies on OCI and container runtimes
+
+Cgroups v2 requires changes to the OCI runtime specs. Changes to the
+OCI specs are out of the scope for this proposal.
+
+### Current status of dependencies
+
+- CRI-O+crun: support cgroups v2
+
+- runc support for cgroups v2 is work in progress
+
+- containerd: [https://github.com/containerd/containerd/pull/3799](https://github.com/containerd/containerd/pull/3799)
+
+- Moby: [https://github.com/moby/moby/pull/40174](https://github.com/moby/moby/pull/40174)
+
+- OCI runtime spec: TODO
+
+## Current cgroups usage and the equivalent in cgroups v2
+
+|Kubernetes cgroups v1|Kubernetes cgroups v2 behavior|
+|---|---|
+|CPU stats for Horizontal Pod Autoscaling|No .percpu cpuacct stats.  Check cadvisor.|
```

Comment:
Does this have to depend on CAdvisor? I thought we were trying to depend on the stats API instead of requiring cadvisor.

Reply:
is dropping CAdvisor tracked somewhere?  I've added only because I saw that it is still used.

Also the changes requires for CAdvisor were already merged upstream.  Another iteration might be required once hugetlb is supported though.